src/corner-kick/src/service/RosConnector.ts

```diff
@@ -0,0 +1,132 @@
+    public static get Instance() {
+        return this.instance || (this.instance = new this());
+    }
+
+    private constructor() {
+        // Wait a little bit before connecting to ROS. It can sometimes take time
+        // to initialize.
+        setTimeout(() => {
+            this.ros = new ROSLib.Ros({
+                url: ROS_URL,
+            });
+
+            // TODO: handle connection errors
+
+            // We are connected, subscribe to ROS
+            this.subscribeToROS();
+        }, 1000);
+    }
+
+    /**
+     * Subscribe to a given topic. When new messages are received on the topic, the callback will be called.
+     * @param topic the topic to subscribe to
+     * @param callback the callback function to call when a new topic is received
+     */
+    public subscribeToTopic = (topic: string, callback: (topic: any) => void) => {
+        // We initialize a new array if this is the first callback for a given topic
+        if(this.topicsSubscriptions[topic] === undefined) {
+            this.topicsSubscriptions[topic] = [callback];
+        } else {
+            this.topicsSubscriptions[topic].push(callback);
+        }
+    }
+
+    /**
+     * Unsubscribe from a given topic.
+     * @param topic the topic to unsubscribe from
+     * @param callback the callback function to remove
+     */
+    public unsubscribeToTopic = (topic: string, callback: (topic: any) => void) => {
+        if(this.topicsSubscriptions[topic] !== undefined) {
+            const index = this.topicsSubscriptions[topic].indexOf(callback);
+            this.topicsSubscriptions[topic].splice(index, 1);
+        }
+    }
+
+    public publish = (name: string, messageType: string, message: any) => {
+        const topic = new ROSLib.Topic({
+            messageType,
+            name,
+            ros: this.ros,
+        });
+
+        topic.publish(message);
+    }
+
+    /**
+     * Internal method to subscribe to all ROS topics. This is the method that actually "subscribes" in a ROS sense. We do this to create an abstraction
+     * layer between ROS and the rest of our application
+     */
+    private subscribeToROS = () => {
+        // Request ROS to give us a list of all topics.
+        this.ros.getTopics(this.onTopicsReceived);
+
+        // Every 10 seconds, ask for a new list from ROS. If new topics are added during runtime, we will pick them up.
+        setTimeout(this.subscribeToROS, 10000);
+    }
+
+    /**
+     * This internal method subscribes (in a ROS sense) to every topic found. When a new message is received, it calls
+     * onMessageReceived, which does the actual act of notifying the rest of the application.
+     */
+    private onTopicsReceived = (response: {topics: string[], types: string[]}) => {
+        const newTopics: string[] = [];
+
+        response.topics.forEach((element, index) => {
+            // We only subscribe to topics we have not seen yet
+            if(this.topics.indexOf(element) === -1) {
```

Comment:
Could this be `if (this.topics.includes(element)){` instead? https://stackoverflow.com/questions/42790602/how-do-i-check-whether-an-array-contains-a-string-in-typescript